---
output: github_document
---

```{r, echo = FALSE, message = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  rows.print=5
)
options(tidyverse.quiet = TRUE)
```

# kentr 

![R-CMD-check](https://github.com/kriemo/kentr/workflows/R-CMD-check/badge.svg)


A R package with a collection of functions for working with sequence data in R. Includes the [`ssw`](https://github.com/mengyao/Complete-Striped-Smith-Waterman-Library) Smith-Waterman  alignment C/C++ library.

```r
#install.packages('devtools')
devtools::install_github('kriemo/kentr')
```

## Basic Usage

### Extract sequences from fasta:
```{r}
library(kentr)
# generate data.frame (or tibble) with bed coordinates

df <- data.frame(chrom = "chr1",
                 start = 20000,
                 end = 20025)

df

# path to fasta file (with fai samtools faidx index)
fa_path <- system.file("extdata", "test.fasta", package = "kentr")

get_sequences(df, fa_path)
```

### Compute Hamming distances:

```{r}
fa_path <- system.file("extdata", "test.fasta", package = "kentr")
df <- data.frame(chrom = "chr1",
                 start = c(20000, 25000),
                 end = c(20025, 25025))

df

seqs <- get_sequences(df, fa_path)

seq1 <- seqs[1, "seq"]
seq2 <- seqs[2, "seq"]
get_hamming(seq1, seq2)
```


### Count kmers in sequences

```{r}
library(tidyverse, warn.conflicts = FALSE)

fa_path <- system.file("extdata", "test.fasta", package = "kentr")
df <- data.frame(chrom = "chr1",
                 start = c(20000, 25000),
                 end = c(20500, 25500))

df

seqs <- get_sequences(df, fa_path)

seqs

# returns list with a dataframe of kmers and counts for each sequence
get_kmers(seqs$seq, n = 2)

# use with dplyr tibbles 

seqs <- as_tibble(seqs)
seqs

kmers <- mutate(seqs, 
                kmers = get_kmers(seq)) %>% 
                select(header, kmers) 
                
kmers

unnest(kmers)
```

### Perform Smith-Waterman alignment

Uses the [Complete-Striped-Smith-Waterman-Library](https://github.com/mengyao/Complete-Striped-Smith-Waterman-Library) to perform alignment between a query sequence and vector of reference sequences. 
```{r}
query_seq <- seqs$seq[1]
ref_seqs <- seqs$seq

get_sw(query_seq, ref_seqs)
```


### Extract bam alignments as a data.frame

```{r}

bam_file <- system.file("extdata",
                        "small_sorted.bam", 
                        package = "kentr")
res <- bam_to_df(bam_file)
head(res)

res <- bam_to_df(bam_file,
                 region = "chr1:3e6-3.1e6",
                 tags = c("XS:A", "AS:i"))
res
```

### Extract tabix data as a data.frame

```{r}
tbx_file <- system.file("extdata", "tabix.bed.gz", package = "kentr")
res <- tabix_to_df(tbx_file)
head(res)

res <- tabix_to_df(tbx_file,
                    region = "chr1:3e6-3.1e6")
res
```


### Find motif matches genome-wide

```{r}
fa_path <- system.file("extdata", "test.fasta", package = "kentr")
query_seq <- "AATAAA[GTC]"

matches <- read_genome_seq(fa_path) %>% 
  find_motifs(query_seq) %>% 
  head()

matches

get_sequences(matches, fa_path)
```
