# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

read_bam_tags <- function(bampath, tag_ids, tag_types, region = ".") {
    .Call(`_kentr_read_bam_tags`, bampath, tag_ids, tag_types, region)
}

read_bam <- function(bampath, region, tag_ids, tag_types) {
    .Call(`_kentr_read_bam`, bampath, region, tag_ids, tag_types)
}

#' Fetch DNA sequence from an indexed fasta file
#' @param df dataframe wtih columns chrom start and end
#' @param fapath path to indexed fasta file
#' @export
getSeq <- function(df, fapath) {
    .Call(`_kentr_getSeq`, df, fapath)
}

#' Reverse complement a vector of DNA sequences
#' @param vec character vector of DNA [ATCGN]
#' @export
revComp <- function(vec) {
    .Call(`_kentr_revComp`, vec)
}

#' Compare a vector of strings to a vector of strings and report hamming distance
#' @param bcs_to_test characater vector to get hamming distance
#' @param all_bcs character vector to compare
#' @export
get_hamming_pairs <- function(bcs_to_test, all_bcs) {
    .Call(`_kentr_get_hamming_pairs`, bcs_to_test, all_bcs)
}

#' Compare a string to a vector of strings and report hamming distance
#' @param bc_to_test String to get hamming distance
#' @param all_bcs character vector to compare
#' @export
get_hamming <- function(bc_to_test, all_bcs) {
    .Call(`_kentr_get_hamming`, bc_to_test, all_bcs)
}

#' Compare a string to a vector of strings report mismatch position
#' @param bc_to_test String to check
#' @param all_bcs character vector to compare
#' @export
find_mismatch <- function(bc_to_test, all_bcs) {
    .Call(`_kentr_find_mismatch`, bc_to_test, all_bcs)
}

#' get kmer-counts for character vector of sequences
#' @param seqs character vector of sequences
#' @param n kmer size
#' @export
get_kmers <- function(seqs, n = 2L) {
    .Call(`_kentr_get_kmers`, seqs, n)
}

mw_test_impl <- function(df, idx1, idx2) {
    .Call(`_kentr_mw_test_impl`, df, idx1, idx2)
}

#' Compute complete striped smith waterman alignment between two dna strings (ATCG)
#' @description Alignments will be computed for both forward and reverse complement. Forward
#' alignments will be indicated with strand "+", whereas reverse complement alignments will
#' be reported as "-". The best alignment is reported
#' @param query_seq query sequence as character vector of length one
#' @param ref_seqs reference sequences as a character vector
#' @export
get_sw <- function(query_seq, ref_seqs) {
    .Call(`_kentr_get_sw`, query_seq, ref_seqs)
}

